src/astra_driver.cpp

```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
   }
 }
 
+bool AstraDriver::validateGrabImagesGoal(const camera_control_msgs::GrabImagesGoalConstPtr &goal)
+{
+  if (goal->exposure_given and goal->gain_given)
+  {
+    if (goal->exposure_times.size() != goal->gain_values.size())
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+void AstraDriver::colorGrabImagesCallback(const camera_control_msgs::GrabImagesGoalConstPtr &goal)
+{
+  // Only the auto exposure, exposure and gain parameters of the goal are used
+  if (goal->exposure_given or goal->gain_given)
+  {
+    ROS_WARN_ONCE("Received goal with gamma or brightness given. These values "
+                  "are not supported and will be ignored!");
+  }
+
+  imageConnectCb(true); // Connect (in case there are no subscribers)
+  CameraParameters old_parameters = getCameraParameters(); // Backup parameters
+
+  camera_control_msgs::GrabImagesResult result;
+
+  if (not validateGrabImagesGoal(goal))
+  {
+    color_action_server_->setAborted(result, "Invalid goal");
+    return;
+  }
+
+  device_->setAutoExposure(goal->exposure_auto);
+
+  // TODO Check parameter units and transform parameters if necessary (orbbec
+  // doesn't provide any info on this)
+
+  actionlib_msgs::GoalStatus goal_status;
+  goal_status.status = actionlib_msgs::GoalStatus::SUCCEEDED;
+  camera_control_msgs::GrabImagesFeedback feedback;
+  size_t num_imgs =
+      std::max((size_t)1, std::max(goal->exposure_times.size(), goal->gain_values.size()));
+  for (size_t i = 0; i < num_imgs; ++i)
+  {
+    if (color_action_server_->isPreemptRequested())
+    {
+      goal_status.status = actionlib_msgs::GoalStatus::PREEMPTED;
+      break;
+    }
+
+    // Set parameters
+    if (goal->exposure_given)
+    {
+      device_->setIRExposure(static_cast<int>(goal->exposure_times[i]));
+    }
+    if (goal->gain_given)
+    {
+      device_->setIRGain(static_cast<int>(goal->gain_values[i]));
+    }
+
+    // Get image
+    boost::mutex::scoped_lock lock(color_syncer_.mutex);
+    color_syncer_.reset(true);
+    bool got_image = color_syncer_.cv.wait_for(
+        lock, boost::chrono::seconds(1), boost::bind(&ActionImageSyncer::ready, &color_syncer_));
+    if (not got_image)
+    {
+      ROS_WARN("Failed to grab images; color stream is unresponsive");
+      break;
+    }
+    result.images.push_back(*color_syncer_.image);
```

Comment:
the problem with this code is that your image is a shared ptr. without resetting it, you will hold it also after finishing the service request

Reply:
Done [here](https://github.com/magazino/ros_astra_camera/blob/add_grab_image_action/src/astra_driver.cpp#L686)